internal/controller/management/upgrade/freight.go

```diff
@@ -0,0 +1,132 @@
+func SetupFreightReconcilerWithManager(mgr manager.Manager) error {
+	_, err := ctrl.NewControllerManagedBy(mgr).
+		For(&kargoapi.Freight{}).
+		WithEventFilter(
+			predicate.Funcs{
+				DeleteFunc: func(event.DeleteEvent) bool {
+					return false
+				},
+			},
+		).
+		WithEventFilter(predicate.NewPredicateFuncs(func(object client.Object) bool {
+			freight, ok := object.(*kargoapi.Freight)
+			if !ok {
+				return false
+			}
+
+			// Ignore Freight that are already v0.8 compatible.
+			if _, ok = freight.Labels[kargoapi.V08CompatibilityLabelKey]; ok {
+				return false
+			}
+
+			// Ignore Freight that have an origin set.
+			if freight.Origin.Kind != "" && freight.Origin.Name != "" {
+				return false
+			}
+
+			// Ignore Freight that are missing the Warehouse field.
+			return freight.Warehouse != "" // nolint: staticcheck
+		})).
+		Build(&freightReconciler{
+			client: mgr.GetClient(),
+		})
+	return err
+}
+
+// Reconcile is part of the main Kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (f *freightReconciler) Reconcile(
+	ctx context.Context,
+	req ctrl.Request,
+) (ctrl.Result, error) {
+	logger := logging.LoggerFromContext(ctx).WithValues(
+		"namespace", req.NamespacedName.Namespace,
+		"freight", req.NamespacedName.Name,
+	)
+	logger.Debug("reconciling Freight")
+
+	// Check if the freights.kargo.akuity.io CRD has the origin field.
+	// If it does not, we need to wait for the CRD to be updated.
+	var freightCRD extv1.CustomResourceDefinition
+	if err := f.client.Get(
+		ctx,
+		types.NamespacedName{
+			Name: "freights.kargo.akuity.io",
+		},
+		&freightCRD,
+	); err != nil {
+		return ctrl.Result{}, err
+	}
+	if _, hasOriginField := freightCRD.Spec.Versions[0].Schema.OpenAPIV3Schema.
+		Properties["origin"]; !hasOriginField {
+		return ctrl.Result{}, errors.New(
+			"freights.kargo.akuity.io does not have an origin field: waiting for update",
+		)
+	}
+
+	// Find the Freight.
+	freight, err := kargoapi.GetFreight(ctx, f.client, req.NamespacedName)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if freight == nil {
+		// Ignore if not found. This can happen if the Freight was deleted after the
+		// current reconciliation request was issued.
+		return ctrl.Result{}, nil // Do not requeue
+	}
+
+	// Check if the Freight is already v0.8.x compatible.
+	if freight.Origin.Kind != "" && freight.Origin.Name != "" {
+		logger.Debug("Freight is already v0.8 compatible")
+		return ctrl.Result{}, nil
+	}
+
+	// If the Warehouse field is not set, we can't migrate the Freight.
+	if freight.Warehouse == "" { // nolint: staticcheck
+		logger.Debug("Freight is missing Warehouse")
+		return ctrl.Result{}, nil
+	}
+
+	// Migrate the Warehouse field to the Origin field.
+	freight.Origin.Kind = kargoapi.FreightOriginKindWarehouse
+	freight.Origin.Name = freight.Warehouse // nolint: staticcheck
+	freight.Warehouse = ""                  // nolint: staticcheck
```

Comment:
@hiddeco I am thinking we shouldn't blank out `freight.Warehouse`. If some shards are still running older controllers, keeping the value in this field would allow them to continue functioning correctly until they are upgraded. wdyt?

Reply:
Am fine with this if we ensure some validation is in place that it is in sync with `.origin.name` if present.

Reply:
That's a good idea. I'll handle that.